### PR TITLE
[6.3] Cover virt-who hypervisor search and subscriptions (BZs: 1246554, 1382090, 1426403)

### PR DESCRIPTION
**Pull request by ldjebran**
qe_test_coverage
* https://bugzilla.redhat.com/show_bug.cgi?id=1246554
* https://bugzilla.redhat.com/show_bug.cgi?id=1426403
* https://bugzilla.redhat.com/show_bug.cgi?id=1382090

Found bugs: 
* https://bugzilla.redhat.com/show_bug.cgi?id=1506549  
* https://bugzilla.redhat.com/show_bug.cgi?id=1506636

Test results: 
* sat-6.3 snap 20 (run without skipping, one test failed as per bug description as bug not fixed yet)

```console
(sat-6.3.0) dlezz@elysion:~/projects/robottelo-fork$ pytest -v tests/foreman/ui/test_virt_who_config.py::VirtWhoConfigDeployedTestCase
================================================= test session starts ==================================================
platform linux2 -- Python 2.7.13, pytest-3.1.3, py-1.4.34, pluggy-0.4.0 -- /home/dlezz/.pyenv/versions/sat-6.3.0/bin/python2.7
cachedir: .cache
rootdir: /home/dlezz/projects/robottelo-fork, inifile:
plugins: xdist-1.20.1, services-1.2.1, mock-1.6.2, forked-0.2, cov-2.5.1
collected 3 items 
2017-10-26 14:11:58 - conftest - DEBUG - Deselect of WONTFIX BZs is disabled in settings


tests/foreman/ui/test_virt_who_config.py::VirtWhoConfigDeployedTestCase::test_positive_open_hypervisor_contenthost_from_subscription <- ../../.pyenv/versions/sat-6.3.0/lib/python2.7/site-packages/robozilla/decorators/__init__.py FAILED
tests/foreman/ui/test_virt_who_config.py::VirtWhoConfigDeployedTestCase::test_positive_search_by_hypervisor PASSED
tests/foreman/ui/test_virt_who_config.py::VirtWhoConfigDeployedTestCase::test_positive_vdc_subscription_contenthost_association PASSED
======================================== 1 failed, 2 passed in 3520.98 seconds =========================================
```

* sat-6.3 snap 21

```console
(sat-6.3.0) dlezz@elysion:~/projects/robottelo-fork$ pytest -v tests/foreman/ui/test_virt_who_config.py::VirtWhoConfigDeployedTestCase
================================================= test session starts ==================================================
platform linux2 -- Python 2.7.13, pytest-3.1.3, py-1.4.34, pluggy-0.4.0 -- /home/dlezz/.pyenv/versions/sat-6.3.0/bin/python2.7
cachedir: .cache
rootdir: /home/dlezz/projects/robottelo-fork, inifile:
plugins: xdist-1.20.1, services-1.2.1, mock-1.6.2, forked-0.2, cov-2.5.1
collected 3 items 
2017-10-26 16:28:18 - conftest - DEBUG - Deselect of WONTFIX BZs is disabled in settings


tests/foreman/ui/test_virt_who_config.py::VirtWhoConfigDeployedTestCase::test_positive_open_hypervisor_contenthost_from_subscription <- ../../.pyenv/versions/sat-6.3.0/lib/python2.7/site-packages/robozilla/decorators/__init__.py SKIPPED
tests/foreman/ui/test_virt_who_config.py::VirtWhoConfigDeployedTestCase::test_positive_search_by_hypervisor PASSED
tests/foreman/ui/test_virt_who_config.py::VirtWhoConfigDeployedTestCase::test_positive_vdc_subscription_contenthost_association <- ../../.pyenv/versions/sat-6.3.0/lib/python2.7/site-packages/robozilla/decorators/__init__.py SKIPPED

======================================== 1 passed, 2 skipped in 1967.56 seconds ========================================
```

### CODE REVIEW

**Comment by ldjebran** [comment]
put bug https://bugzilla.redhat.com/show_bug.cgi?id=1506549 as duplicate of 
https://bugzilla.redhat.com/show_bug.cgi?id=1495271
put DO NOT MERGE to change skipt_if_open

**Comment by ldjebran** [comment]
```console
(sat-6.3.0) dlezz@elysion:~/projects/robottelo-fork$ pytest -v tests/foreman/ui/test_virt_who_config.py::VirtWhoConfigDeployedTestCase
================================================= test session starts ==================================================
platform linux2 -- Python 2.7.13, pytest-3.1.3, py-1.4.34, pluggy-0.4.0 -- /home/dlezz/.pyenv/versions/sat-6.3.0/bin/python2.7
cachedir: .cache
rootdir: /home/dlezz/projects/robottelo-fork, inifile:
plugins: xdist-1.20.1, services-1.2.1, mock-1.6.2, forked-0.2, cov-2.5.1
collected 3 items 
2017-10-27 09:56:17 - conftest - DEBUG - Deselect of WONTFIX BZs is disabled in settings


tests/foreman/ui/test_virt_who_config.py::VirtWhoConfigDeployedTestCase::test_positive_open_hypervisor_contenthost_from_subscription <- ../../.pyenv/versions/sat-6.3.0/lib/python2.7/site-packages/robozilla/decorators/__init__.py SKIPPED
tests/foreman/ui/test_virt_who_config.py::VirtWhoConfigDeployedTestCase::test_positive_search_by_hypervisor PASSED
tests/foreman/ui/test_virt_who_config.py::VirtWhoConfigDeployedTestCase::test_positive_vdc_subscription_contenthost_association <- ../../.pyenv/versions/sat-6.3.0/lib/python2.7/site-packages/robozilla/decorators/__init__.py SKIPPED

======================================== 1 passed, 2 skipped in 1839.48 seconds ========================================
```

File: tests/foreman/ui/test_virt_who_config.py [hunk 1]
```diff
@@ -205,3 +229,176 @@ def test_positive_lastest_out_of_date(self):
...
+        :expectedresults: the hypervisor search work properly
+
+        :BZ: 1246554, 1495271
+
+        :CaseLevel: System
+        """
+        with Session(self) as session:
+            set_context(session, org=self.session_org.name)
+            is_hypervisor_hosts = {
+                'true': self.virt_who_hypervisor_host.name,
```
Comment (oshtaier) [inline tests/foreman/ui/test_virt_who_config.py]: why do you use string keys and then use that ugly if condition? I think we can use boolean ones

File: tests/foreman/ui/test_virt_who_config.py [hunk 2]
```diff
@@ -205,3 +229,176 @@ def test_positive_lastest_out_of_date(self):
...
+
+    @classmethod
+    def tearDownClass(cls):
+        """destroy the virt-who virtual machine"""
+        cls.virt_who_vm.destroy()
+        super(VirtWhoConfigDeployedTestCase, cls).tearDownClass()
+
+    @tier4
+    def test_positive_search_by_hypervisor(self):
+        """Able to found hypervisor or not hypervisor content host by searching
```
Comment (oshtaier) [inline tests/foreman/ui/test_virt_who_config.py]: `to find`

File: tests/foreman/ui/test_virt_who_config.py (hunk 1 above)
Comment (abalakh) [reply tests/foreman/ui/test_virt_who_config.py]: and `str(not is_hypervisor_hosts).lower()` correspondingly

**Comment by ldjebran** [comment]
@abalakh @oshtaier all comments addressed
```console
(sat-6.3.0) dlezz@elysion:~/projects/robottelo-fork$ pytest -v tests/foreman/ui/test_virt_who_config.py::VirtWhoConfigDeployedTestCase
=================================================================================================== test session starts ====================================================================================================
platform linux2 -- Python 2.7.13, pytest-3.1.3, py-1.4.34, pluggy-0.4.0 -- /home/dlezz/.pyenv/versions/sat-6.3.0/bin/python2.7
cachedir: .cache
rootdir: /home/dlezz/projects/robottelo-fork, inifile:
plugins: xdist-1.20.1, services-1.2.1, mock-1.6.2, forked-0.2, cov-2.5.1
collected 3 items 
2017-10-27 16:33:13 - conftest - DEBUG - Deselect of WONTFIX BZs is disabled in settings


tests/foreman/ui/test_virt_who_config.py::VirtWhoConfigDeployedTestCase::test_positive_open_hypervisor_contenthost_from_subscription <- ../../.pyenv/versions/sat-6.3.0/lib/python2.7/site-packages/robozilla/decorators/__init__.py SKIPPED
tests/foreman/ui/test_virt_who_config.py::VirtWhoConfigDeployedTestCase::test_positive_search_by_hypervisor PASSED
tests/foreman/ui/test_virt_who_config.py::VirtWhoConfigDeployedTestCase::test_positive_vdc_subscription_contenthost_association <- ../../.pyenv/versions/sat-6.3.0/lib/python2.7/site-packages/robozilla/decorators/__init__.py SKIPPED

========================================================================================== 1 passed, 2 skipped in 1788.18 seconds ==========================================================================================
```

File: tests/foreman/ui/test_virt_who_config.py [hunk 3]
```diff
@@ -205,3 +229,173 @@ def test_positive_lastest_out_of_date(self):
...
+        """
+        with Session(self) as session:
+            set_context(session, org=self.session_org.name)
+            self.assertIsNotNone(
+                session.contenthost.search(self.virt_who_hypervisor_host.name))
+            session.subscriptions.search_and_click(VDC_SUBSCRIPTION_NAME)
+            session.subscriptions.click(tab_locators['subs.sub.associations'])
+            session.subscriptions.click(
+                tab_locators['subs.sub.associations.hosts'])
+            session.subscriptions.wait_for_ajax()
```
Comment (abalakh) [inline tests/foreman/ui/test_virt_who_config.py]: Most likely this line is redundant, as Base.click() calls `wait_for_ajax()` after actual clicking

Comment (ldjebran) [reply tests/foreman/ui/test_virt_who_config.py]: @abalakh wanted to make sure, that everything is OK before calling find_elements, as that UI is not that logical as it appear and the list do not appear immediately after clicking (after all functions initiated by the click are resumed), perhaps an other call is initiated by an other function.

Comment (ldjebran) [reply tests/foreman/ui/test_virt_who_config.py]: I have no chance now to test without wait_for_ajax (tested on earlier version 2 weeks ago), what I can say that with current version the find elements work with a ratio of 100% without it 50%

File: tests/foreman/ui/test_virt_who_config.py [hunk 4]
```diff
@@ -205,3 +229,172 @@ def test_positive_lastest_out_of_date(self):
...
+class VirtWhoConfigDeployedTestCase(UITestCase):
+    """Implements Virt-who-config UI tests with virt-who config deployed"""
+
+    @classmethod
+    def set_session_org(cls):
+        """Create global test case organization"""
+        cls.session_org = entities.Organization().create()
+
+    @classmethod
+    @skip_if_not_set('clients', 'fake_manifest')
```
Comment (oshtaier) [inline tests/foreman/ui/test_virt_who_config.py]: +`compute_resources`

**Comment by ldjebran** [comment]
```shell
(sat-6.3.0) dlezz@elysion:~/projects/robottelo-fork$ pytest -v tests/foreman/ui/test_virt_who_config.py::VirtWhoConfigDeployedTestCase
================================================= test session starts ==================================================
platform linux2 -- Python 2.7.13, pytest-3.1.3, py-1.4.34, pluggy-0.4.0 -- /home/dlezz/.pyenv/versions/sat-6.3.0/bin/python2.7
cachedir: .cache
rootdir: /home/dlezz/projects/robottelo-fork, inifile:
plugins: xdist-1.20.1, services-1.2.1, mock-1.6.2, forked-0.2, cov-2.5.1
collected 3 items 
2017-11-08 16:54:20 - conftest - DEBUG - Deselect of WONTFIX BZs is disabled in settings


tests/foreman/ui/test_virt_who_config.py::VirtWhoConfigDeployedTestCase::test_positive_open_hypervisor_contenthost_from_subscription <- ../../.pyenv/versions/sat-6.3.0/lib/python2.7/site-packages/robozilla/decorators/__init__.py SKIPPED
tests/foreman/ui/test_virt_who_config.py::VirtWhoConfigDeployedTestCase::test_positive_search_by_hypervisor PASSED
tests/foreman/ui/test_virt_who_config.py::VirtWhoConfigDeployedTestCase::test_positive_vdc_subscription_contenthost_association <- ../../.pyenv/versions/sat-6.3.0/lib/python2.7/site-packages/robozilla/decorators/__init__.py SKIPPED

======================================== 1 passed, 2 skipped in 1705.12 seconds ========================================
```

**Comment by ldjebran** [comment]
@oshtaier all comments addressed

**Pull request merged**